File: packages/domain/src/rules/ids.rs
use thiserror::Error;

const TRACE_ID_LEN: usize = 32;
const SPAN_ID_LEN: usize = 16;

#[derive(Debug, Error)]
pub enum IdError {
    #[error("invalid trace id: {0}")]
    TraceId(String),
    #[error("invalid span id: {0}")]
    SpanId(String),
}

pub fn normalize_trace_id(input: &str) -> Result<String, IdError> {
    normalize_hex_id(input, TRACE_ID_LEN).map_err(IdError::TraceId)
}

pub fn normalize_span_id(input: &str) -> Result<String, IdError> {
    normalize_hex_id(input, SPAN_ID_LEN).map_err(IdError::SpanId)
}

pub fn normalize_trace_id_bytes(bytes: &[u8]) -> Result<String, IdError> {
    if bytes.len() != 16 {
        return Err(IdError::TraceId(format!("expected 16 bytes, got {}", bytes.len())));
    }
    if bytes.iter().all(|b| *b == 0) {
        return Err(IdError::TraceId("all-zero trace id".into()));
    }
    Ok(hex_encode(bytes))
}

pub fn normalize_span_id_bytes(bytes: &[u8]) -> Result<String, IdError> {
    if bytes.len() != 8 {
        return Err(IdError::SpanId(format!("expected 8 bytes, got {}", bytes.len())));
    }
    if bytes.iter().all(|b| *b == 0) {
        return Err(IdError::SpanId("all-zero span id".into()));
    }
    Ok(hex_encode(bytes))
}

pub fn optional_trace_id(input: &str) -> Option<String> {
    normalize_trace_id(input).ok()
}

pub fn optional_span_id(input: &str) -> Option<String> {
    normalize_span_id(input).ok()
}

pub fn optional_trace_id_bytes(bytes: &[u8]) -> Option<String> {
    if bytes.is_empty() {
        return None;
    }
    normalize_trace_id_bytes(bytes).ok()
}

pub fn optional_span_id_bytes(bytes: &[u8]) -> Option<String> {
    if bytes.is_empty() {
        return None;
    }
    normalize_span_id_bytes(bytes).ok()
}

pub fn is_root_parent(parent_span_id: Option<&str>) -> bool {
    match parent_span_id {
        None => true,
        Some(id) => id.is_empty(),
    }
}

fn normalize_hex_id(input: &str, expected_len: usize) -> Result<String, String> {
    let trimmed = input.trim();
    if trimmed.len() != expected_len {
        return Err(format!("expected {expected_len} hex chars, got {}", trimmed.len()));
    }
    if !trimmed.chars().all(|c| c.is_ascii_hexdigit()) {
        return Err("non-hex character".into());
    }
    let lower = trimmed.to_ascii_lowercase();
    if lower.chars().all(|c| c == '0') {
        return Err("all-zero id".into());
    }
    Ok(lower)
}

fn hex_encode(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn accepts_uppercase_hex() {
        assert_eq!(
            normalize_trace_id("5B8EFFF798038103D269B633813FC60C").unwrap(),
            "5b8efff798038103d269b633813fc60c"
        );
    }

    #[test]
    fn rejects_all_zero() {
        assert!(normalize_span_id("0000000000000000").is_err());
    }
}

File: apps/api/src/lib/ids.ts
const TRACE_ID_LEN = 32
const SPAN_ID_LEN = 16

export class IdError extends Error {
  constructor(message: string) {
    super(message)
    this.name = "IdError"
  }
}

function hexEncode(bytes: Uint8Array): string {
  return Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("")
}

function normalizeHexId(input: string, expectedLen: number, kind: "trace" | "span"): string {
  const trimmed = input.trim()
  if (trimmed.length !== expectedLen) {
    throw new IdError(
      `invalid ${kind} id: expected ${expectedLen} hex chars, got ${trimmed.length}`,
    )
  }
  if (![...trimmed].every((c) => /[0-9a-fA-F]/.test(c))) {
    throw new IdError(`invalid ${kind} id: non-hex character`)
  }
  const lower = trimmed.toLowerCase()
  if ([...lower].every((c) => c === "0")) {
    throw new IdError(`invalid ${kind} id: all-zero id`)
  }
  return lower
}

export function normalizeTraceId(input: string): string {
  return normalizeHexId(input, TRACE_ID_LEN, "trace")
}

export function normalizeSpanId(input: string): string {
  return normalizeHexId(input, SPAN_ID_LEN, "span")
}

export function normalizeTraceIdBytes(bytes: Uint8Array): string {
  if (bytes.length !== 16) {
    throw new IdError(`invalid trace id: expected 16 bytes, got ${bytes.length}`)
  }
  if (bytes.every((b) => b === 0)) {
    throw new IdError("invalid trace id: all-zero trace id")
  }
  return hexEncode(bytes)
}

export function normalizeSpanIdBytes(bytes: Uint8Array): string {
  if (bytes.length !== 8) {
    throw new IdError(`invalid span id: expected 8 bytes, got ${bytes.length}`)
  }
  if (bytes.every((b) => b === 0)) {
    throw new IdError("invalid span id: all-zero span id")
  }
  return hexEncode(bytes)
}

export function optionalTraceId(input: string): string | undefined {
  try {
    return normalizeTraceId(input)
  } catch {
    return undefined
  }
}

export function optionalSpanId(input: string): string | undefined {
  try {
    return normalizeSpanId(input)
  } catch {
    return undefined
  }
}

export function optionalTraceIdBytes(bytes: Uint8Array): string | undefined {
  if (bytes.length === 0) return undefined
  try {
    return normalizeTraceIdBytes(bytes)
  } catch {
    return undefined
  }
}

export function optionalSpanIdBytes(bytes: Uint8Array): string | undefined {
  if (bytes.length === 0) return undefined
  try {
    return normalizeSpanIdBytes(bytes)
  } catch {
    return undefined
  }
}

export function isRootParent(parentSpanId: string | undefined | null): boolean {
  return parentSpanId == null || parentSpanId === ""
}

export function parseOtlpId(
  value: unknown,
  kind: "trace" | "span",
): string {
  if (typeof value === "string") {
    return kind === "trace" ? normalizeTraceId(value) : normalizeSpanId(value)
  }
  if (value instanceof Uint8Array) {
    return kind === "trace" ? normalizeTraceIdBytes(value) : normalizeSpanIdBytes(value)
  }
  if (Buffer.isBuffer(value)) {
    return kind === "trace"
      ? normalizeTraceIdBytes(value)
      : normalizeSpanIdBytes(value)
  }
  throw new IdError(`invalid ${kind} id`)
}

export function optionalOtlpId(
  value: unknown,
  kind: "trace" | "span",
): string | undefined {
  if (value == null || value === "") return undefined
  if (value instanceof Uint8Array && value.length === 0) return undefined
  try {
    return parseOtlpId(value, kind)
  } catch {
    return undefined
  }
}
